dev/com.ibm.ws.microprofile.faulttolerance.2.0/src/com/ibm/ws/microprofile/faulttolerance20/state/impl/AsyncBulkheadStateImpl.java

```diff
@@ -0,0 +1,106 @@
+/*******************************************************************************
+ * Copyright (c) 2018 IBM Corporation and others.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *     IBM Corporation - initial API and implementation
+ *******************************************************************************/
+package com.ibm.ws.microprofile.faulttolerance20.state.impl;
+
+import java.util.concurrent.ConcurrentLinkedQueue;
+import java.util.concurrent.Future;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.Semaphore;
+
+import com.ibm.ws.microprofile.faulttolerance.spi.BulkheadPolicy;
+import com.ibm.ws.microprofile.faulttolerance20.state.AsyncBulkheadState;
+
+public class AsyncBulkheadStateImpl implements AsyncBulkheadState {
+
+    private final ScheduledExecutorService executorService;
+    private final ConcurrentLinkedQueue<ExecutionReferenceImpl> queue;
+    private final Semaphore executionSemaphore;
+    private final Semaphore queuingSemaphore;
+
+    public AsyncBulkheadStateImpl(ScheduledExecutorService executorService, BulkheadPolicy policy) {
+        this.executorService = executorService;
+        this.executionSemaphore = new Semaphore(policy.getMaxThreads());
+        this.queuingSemaphore = new Semaphore(policy.getQueueSize());
+        this.queue = new ConcurrentLinkedQueue<>();
+    }
+
+    @Override
+    public ExecutionReference submit(Runnable runnable) {
+        ExecutionReferenceImpl execution = new ExecutionReferenceImpl(runnable);
+
+        if (executionSemaphore.tryAcquire()) {
+            enqueueExecution(execution);
+            execution.accepted = true;
+        } else if (queuingSemaphore.tryAcquire()) {
+            queue.offer(execution);
+            execution.accepted = true;
+        }
+
+        return execution;
+    }
+
+    private void enqueueExecution(ExecutionReferenceImpl execution) {
+        Future<?> future = executorService.submit(() -> {
+            runRunnable(execution.runnable);
+        });
+        execution.future = future;
+    }
+
+    private void runRunnable(Runnable runnable) {
+        try {
+            runnable.run();
+        } finally {
+            enqueueNext();
+        }
+    }
+
+    private void enqueueNext() {
+        synchronized (this) {
```

Comment:
when is a synchronized block needed or not?

Reply:
Done